File: apps/workbench/frontend/vite/src/api/humanBaseline.ts
import { CASES_PATH, FixtureCaseContext } from "./cases";

export type HumanBaselineArtifactBinding = {
  case_id: string;
  case_version: number;
  research_preview_digest: string;
  analysis_digest: string;
  workpaper_digest: string;
  writer_digest: string;
  artifact_binding_digest: string;
};

export type AnalystBaselineSubmission = {
  strongest_source: string;
  material_limitation: string;
  numeric_verification: string;
  weakest_judgment: string;
  required_modification: string;
  writer_usefulness_score: number;
  writer_usefulness_reason: string;
  time_to_find_source_seconds: number;
  time_to_verify_numeric_seconds: number;
  time_to_identify_weakest_judgment_seconds: number;
  time_to_review_writer_seconds: number;
  repeated_work_count: number;
  blocking_ui_issue: string;
  idempotency_key: string;
};

export type SeniorReviewSubmission = {
  reviewer_ref: string;
  reviewer_role: "senior_analyst" | "domain_reviewer";
  decision: "approve" | "conditional_approve" | "return_for_follow_up";
  research_quality_score: number;
  evidence_quality_score: number;
  senior_reviewability_score: number;
  numeric_reproducibility_confirmed: boolean;
  gap_boundaries_preserved: boolean;
  exact_digest_confirmed: boolean;
  review_comment: string;
  bounded_follow_up: string[];
  idempotency_key: string;
};

export type HumanBaselineSession = {
  schema_version: string;
  session_id: string;
  case_id: string;
  participant_ref: string;
  status: "in_progress" | "analyst_submitted" | "exact_human_senior_review_recorded";
  artifact_binding: HumanBaselineArtifactBinding;
  artifact_binding_digest: string;
  analyst_submission: Omit<AnalystBaselineSubmission, "idempotency_key"> | null;
  senior_review: Omit<SeniorReviewSubmission, "idempotency_key"> | null;
  final_review_digest: string | null;
  started_at: string;
  analyst_submitted_at: string | null;
  senior_reviewed_at: string | null;
  updated_at: string;
  events: Array<{
    event_id: string;
    event_type: string;
    actor_ref: string;
    payload_digest: string;
    created_at: string;
  }>;
  execution_counts: Record<string, number>;
  boundary: string;
};

export type HumanBaselineSessionList = {
  schema_version: string;
  case_id: string;
  sessions: HumanBaselineSession[];
  counts: { session_count: number; completed_review_count: number };
  boundary: string;
};

export class HumanBaselineApiError extends Error {
  readonly code: string;
  readonly statusCode: number;
  readonly detail: unknown;

  constructor(code: string, message: string, statusCode: number, detail?: unknown) {
    super(message);
    this.name = "HumanBaselineApiError";
    this.code = code;
    this.statusCode = statusCode;
    this.detail = detail;
  }
}

const BASELINE_CONTEXT: FixtureCaseContext = {
  tenantId: "fixture_internal",
  projectId: "workbench_internal",
  actorId: "analyst_internal",
  permissions: [
    "case:read",
    "evidence:read",
    "baseline:read",
    "baseline:write",
    "baseline:review",
  ],
};

export class HumanBaselineApiClient {
  constructor(private readonly context: FixtureCaseContext = BASELINE_CONTEXT) {}

  list(caseId: string): Promise<HumanBaselineSessionList> {
    return this.request<HumanBaselineSessionList>(sessionsPath(caseId));
  }

  start(caseId: string, participantRef: string, idempotencyKey: string): Promise<HumanBaselineSession> {
    return this.request<HumanBaselineSession>(sessionsPath(caseId), {
      method: "POST",
      body: JSON.stringify({ participant_ref: participantRef, idempotency_key: idempotencyKey }),
    });
  }

  submitAnalyst(
    caseId: string,
    sessionId: string,
    submission: AnalystBaselineSubmission,
  ): Promise<HumanBaselineSession> {
    return this.request<HumanBaselineSession>(
      `${sessionsPath(caseId)}/${encodeURIComponent(sessionId)}/analyst-submission`,
      { method: "POST", body: JSON.stringify(submission) },
    );
  }

  submitSenior(
    caseId: string,
    sessionId: string,
    submission: SeniorReviewSubmission,
  ): Promise<HumanBaselineSession> {
    return this.request<HumanBaselineSession>(
      `${sessionsPath(caseId)}/${encodeURIComponent(sessionId)}/senior-review`,
      { method: "POST", body: JSON.stringify(submission) },
    );
  }

  private async request<T>(path: string, init: RequestInit = {}): Promise<T> {
    const headers = new Headers(init.headers);
    headers.set("Accept", "application/json");
    if (init.body) headers.set("Content-Type", "application/json");
    headers.set("X-Fin-Case-Tenant", this.context.tenantId);
    headers.set("X-Fin-Case-Project", this.context.projectId);
    headers.set("X-Fin-Case-Actor", this.context.actorId);
    headers.set("X-Fin-Case-Permissions", this.context.permissions.join(","));
    let response: Response;
    try {
      response = await fetch(path, { ...init, headers });
    } catch (error) {
      throw new HumanBaselineApiError("network_unavailable", "Human baseline API is unavailable.", 0, error);
    }
    const text = await response.text();
    const payload = text ? JSON.parse(text) as Record<string, unknown> : {};
    if (!response.ok) {
      const detail = payload.detail as Record<string, unknown> | undefined;
      throw new HumanBaselineApiError(
        String(detail?.reason ?? `http_${response.status}`),
        String(detail?.reason ?? "Human baseline request failed."),
        response.status,
        detail,
      );
    }
    return payload as T;
  }
}

function sessionsPath(caseId: string): string {
  return `${CASES_PATH}/${encodeURIComponent(caseId)}/human-baseline/sessions`;
}

File: apps/workbench/backend/api/v1/human_baseline.py
from __future__ import annotations

from typing import Annotated, Any, Literal

from fastapi import APIRouter, Depends, Header, HTTPException, status
from pydantic import BaseModel, ConfigDict, Field

from ...application.case_service import CasePrincipal
from ...application.human_baseline_service import (
    AnalystBaselineSubmission,
    HumanBaselineService,
    HumanBaselineServiceError,
    SeniorReviewSubmission,
)


class StartHumanBaselineCommand(BaseModel):
    model_config = ConfigDict(extra="forbid")

    participant_ref: str = Field(min_length=1)
    idempotency_key: str = Field(min_length=1)


class SubmitAnalystBaselineCommand(BaseModel):
    model_config = ConfigDict(extra="forbid")

    strongest_source: str = Field(min_length=1)
    material_limitation: str = Field(min_length=1)
    numeric_verification: str = Field(min_length=1)
    weakest_judgment: str = Field(min_length=1)
    required_modification: str = Field(min_length=1)
    writer_usefulness_score: int = Field(ge=1, le=5)
    writer_usefulness_reason: str = Field(min_length=1)
    time_to_find_source_seconds: int = Field(ge=0)
    time_to_verify_numeric_seconds: int = Field(ge=0)
    time_to_identify_weakest_judgment_seconds: int = Field(ge=0)
    time_to_review_writer_seconds: int = Field(ge=0)
    repeated_work_count: int = Field(ge=0)
    blocking_ui_issue: str = ""
    idempotency_key: str = Field(min_length=1)


class SubmitSeniorReviewCommand(BaseModel):
    model_config = ConfigDict(extra="forbid")

    reviewer_ref: str = Field(min_length=1)
    reviewer_role: Literal["senior_analyst", "domain_reviewer"]
    decision: Literal["approve", "conditional_approve", "return_for_follow_up"]
    research_quality_score: int = Field(ge=1, le=5)
    evidence_quality_score: int = Field(ge=1, le=5)
    senior_reviewability_score: int = Field(ge=1, le=5)
    numeric_reproducibility_confirmed: bool
    gap_boundaries_preserved: bool
    exact_digest_confirmed: bool
    review_comment: str = Field(min_length=1)
    bounded_follow_up: list[str] = Field(default_factory=list, max_length=3)
    idempotency_key: str = Field(min_length=1)


def build_human_baseline_router(service: HumanBaselineService) -> APIRouter:
    router = APIRouter(tags=["fin-0.1-human-baseline"])

    @router.get("/cases/{case_id}/human-baseline/sessions")
    def list_sessions(
        case_id: str,
        principal: Annotated[PrincipalHeaders, Depends()],
    ) -> dict[str, Any]:
        return _invoke(lambda: service.list_sessions(case_id, principal.value()))

    @router.post(
        "/cases/{case_id}/human-baseline/sessions",
        status_code=status.HTTP_201_CREATED,
    )
    def start_session(
        case_id: str,
        command: StartHumanBaselineCommand,
        principal: Annotated[PrincipalHeaders, Depends()],
    ) -> dict[str, Any]:
        return _invoke(
            lambda: service.start_session(
                case_id,
                principal.value(),
                participant_ref=command.participant_ref,
                idempotency_key=command.idempotency_key,
            )
        )

    @router.get("/cases/{case_id}/human-baseline/sessions/{session_id}")
    def get_session(
        case_id: str,
        session_id: str,
        principal: Annotated[PrincipalHeaders, Depends()],
    ) -> dict[str, Any]:
        return _invoke(lambda: service.get_session(case_id, session_id, principal.value()))

    @router.post("/cases/{case_id}/human-baseline/sessions/{session_id}/analyst-submission")
    def submit_analyst(
        case_id: str,
        session_id: str,
        command: SubmitAnalystBaselineCommand,
        principal: Annotated[PrincipalHeaders, Depends()],
    ) -> dict[str, Any]:
        return _invoke(
            lambda: service.submit_analyst_baseline(
                case_id,
                session_id,
                principal.value(),
                AnalystBaselineSubmission(**command.model_dump()),
            )
        )

    @router.post("/cases/{case_id}/human-baseline/sessions/{session_id}/senior-review")
    def submit_senior_review(
        case_id: str,
        session_id: str,
        command: SubmitSeniorReviewCommand,
        principal: Annotated[PrincipalHeaders, Depends()],
    ) -> dict[str, Any]:
        payload = command.model_dump()
        payload["bounded_follow_up"] = tuple(payload["bounded_follow_up"])
        return _invoke(
            lambda: service.submit_senior_review(
                case_id,
                session_id,
                principal.value(),
                SeniorReviewSubmission(**payload),
            )
        )

    return router


class PrincipalHeaders:
    def __init__(
        self,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> None:
        self._principal = CasePrincipal(
            tenant_id=(tenant_id or "").strip(),
            project_id=(project_id or "").strip(),
            actor_id=(actor_id or "").strip(),
            permissions=frozenset(
                item.strip() for item in (permissions or "").split(",") if item.strip()
            ),
        )

    def value(self) -> CasePrincipal:
        return self._principal


def _invoke(call: Any) -> dict[str, Any]:
    try:
        return call()
    except HumanBaselineServiceError as exc:
        raise HTTPException(status_code=exc.status_code, detail=exc.detail) from exc
